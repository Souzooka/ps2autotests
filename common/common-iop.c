#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
