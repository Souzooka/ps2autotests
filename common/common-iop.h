#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
