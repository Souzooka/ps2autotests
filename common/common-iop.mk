# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
