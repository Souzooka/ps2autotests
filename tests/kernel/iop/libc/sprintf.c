#include <common-iop.h>
#include <string.h>

//vsprintf is not reliable on a real PS2, we can't use it. We rely on this macro instead.
#define EXECUTE_TEST(output, format, ...) sprintf(output, format, __VA_ARGS__); printf("%s: %s\n", format, output);

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	char buffer[256];

	EXECUTE_TEST(buffer, "%d", 32);
	EXECUTE_TEST(buffer, "%03d", 45);
	EXECUTE_TEST(buffer, "%3d", 12);
	EXECUTE_TEST(buffer, "%3.3d", 87);
	EXECUTE_TEST(buffer, "%+6d", 948);
	EXECUTE_TEST(buffer, "%+6d", -948);
	EXECUTE_TEST(buffer, "%6d", -948);
	EXECUTE_TEST(buffer, "%06d", -948);
	EXECUTE_TEST(buffer, "%+06d", 948);
	EXECUTE_TEST(buffer, "%u", -948);
	EXECUTE_TEST(buffer, "%s%3.3d", "myfile", 0);
	EXECUTE_TEST(buffer, "%x", 0xdeadbeef);
	EXECUTE_TEST(buffer, "%X", 0xDEADBEEF);
	EXECUTE_TEST(buffer, "%08x", 23);
	EXECUTE_TEST(buffer, "%8x", 12);
	EXECUTE_TEST(buffer, "%p", (void*)0x01234567);
	EXECUTE_TEST(buffer, "%c", 'j');
	
	printf("-- TEST END\n");
	return 0;
}
