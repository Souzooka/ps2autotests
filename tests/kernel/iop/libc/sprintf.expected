-- TEST BEGIN
%d: 32
%03d: 045
%3d:  12
%3.3d: 087
%s%3.3d: myfile000
%x: deadbeef
%X: DEADBEEF
%08x: 00000017
%8x:        c
-- TEST END
